seed/php-sdk/imdb/src/Core/SerializableType.php

```diff
@@ -0,0 +1,386 @@
+<?php
+
+namespace Seed\Core;
+
+use DateTime;
+use Exception;
+use InvalidArgumentException;
+use ReflectionProperty;
+use Seed\Core\SerializableType\UnionArrayType;
+
+/**
+ * Abstract class SerializableType
+ *
+ * Provides generic serialization and deserialization methods.
+ */
+abstract class SerializableType
+{
+    /**
+     * Serializes the object to a JSON string.
+     *
+     * @return string
+     */
+    public function toJson(): string
+    {
+        $arrayData = $this->toArray();
+        return json_encode($arrayData, JSON_THROW_ON_ERROR);
+    }
+
+    /**
+     * Serializes the object to an array for JSON encoding.
+     *
+     * @return array
+     */
+    public function toArray(): array
+    {
+        $result = [];
+        $reflectionClass = new \ReflectionClass($this);
+
+        foreach ($reflectionClass->getProperties() as $property) {
+            $jsonKey = self::getJsonKey($property);
+            $value = $property->getValue($this);
+
+            // Handle DateTime properties
+            $dateTypeAttr = $property->getAttributes(DateType::class)[0] ?? null;
+            if ($dateTypeAttr && $value instanceof DateTime) {
+                $dateType = $dateTypeAttr->newInstance()->type;
+                $value = ($dateType === DateType::TYPE_DATE)
+                    ? $value->format('Y-m-d')
+                    : $value->format(DateTime::RFC3339);
```

Comment:
Given that we reference `Y-m-d` and `DateTime::RFC3339` in so many places, it might be worth adding a separate `Core\Constant.php` file:

```php
class Constant {
  public const DateFormat = 'Y-m-d';
  public const DateTimeFormat = DateTime::RFC3339;
}
```

Alternatively, could create a format class if we imagine other formats:

```php
class Format {
  public const Date = 'Y-m-d';
  public const DateTime = DateTime::RFC3339;
}
```

Up to you though - if this feels overkill, I'm down to leave this as-is.

Reply:
Good call, will do!